### PR TITLE
Improve invalid UTF-8 lint by finding the expression initializer

### LINKED ISSUE #115208: `invalid_from_utf8_unchecked` does not fail for include_bytes!
This code should fail if invalid.txt contains invalid UTF-8.
```rs
let str = include_bytes!("../invalid.txt");
let str = unsafe { std::str::from_utf8_unchecked(str) };
```
This should fail as the byte sequence is known at compile time!

### PR DESCRIPTION
**Pull request by Urgau**
This PR introduce a small mechanism to walk up the HIR through bindings, if/else, consts, ... when trying lint on invalid UTF-8.

Fixes https://github.com/rust-lang/rust/issues/115208

### CODE REVIEW

**Comment by Urgau** [comment]
Discussion about the current approach is being [discussed on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Walking.20up.20HIR.20to.20find.20better.20expressions)

r? @WaffleLapkin

**Comment by Urgau** [comment]
Per discussion on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Walking.20up.20HIR.20to.20find.20better.20expressions/near/387523417), I've ported [`clippy_utils::expr_or_init`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.expr_or_init.html) to `LateContext` and adjusted the invalid utf8 lints to use it.

@WaffleLapkin, should be ready for a review.
@rustbot ready

**Comment by WaffleLapkin** [comment]
I have a slight concern about walking HIR to get values, this doesn't sit right with me. HIR doesn't seem like the right IR to do such dataflow-ish analysis. Similarly the error kinds that this lint now looks through seems a bit arbitrary.

That said, this is still a useful lint change, so I'd like another opinion here:
r? @Nilstrieb

As a side note, maybe the lint docs should be changed, so they don't say "literal". Although I'm not sure what would be the best way to describe the lint's new behavior.

**Review by Noratrieb — COMMENTED** [review]
I agree with waffle that the lint docs should be changed and don't have any great ideas either.

While I don't _love_ this code, it looks good enough and is a clear improvement that can also be applied to other lints, so I'm fine with accepting it.

r=me after updating the docs.

**Comment by Urgau** [comment]
I tried to keep it simple, so I've added the phrasing `module bindings` at the end:
```diff
- with an invalid UTF-8 literal.
+ with an invalid UTF-8 literal (modulo bindings).
```

> While I don't _love_ this code, it looks good enough and is a clear improvement that can also be applied to other lints, so I'm fine with accepting it.

I agree with you both, HIR doesn't seems to me either as the right place to do such analysis but I don't know a better place to do it.

> r=me after updating the docs.

Done, but I don't have r+ rights.

@rustbot ready

File: compiler/rustc_lint/src/invalid_from_utf8.rs [hunk 1]
```diff
@@ -11,7 +11,7 @@ use crate::{LateContext, LateLintPass, LintContext};
 declare_lint! {
     /// The `invalid_from_utf8_unchecked` lint checks for calls to
     /// `std::str::from_utf8_unchecked` and `std::str::from_utf8_unchecked_mut`
-    /// with an invalid UTF-8 literal.
+    /// with an invalid UTF-8 literal (modulo bindings).
```
Comment (Nilstrieb) [inline compiler/rustc_lint/src/invalid_from_utf8.rs:14]: To me this reads more like "bindings are not linted". How about "with a known invalid UTF-8 value"?

Comment (Urgau) [reply compiler/rustc_lint/src/invalid_from_utf8.rs]: The "module bindings" was a reference to the literal for which Waffle tough as immediate, so by adding that phrasing I was trying to imply the lint was linting on a literal but that literal might be behind some bindings.

Anyway, changed to your suggestion.

**Review by Noratrieb — APPROVED** [review]

**Pull request merged**